### PR TITLE
Print stack traces on frozen tests in CI

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Instead of existing calls to ps from gnu_parallel, call a new wrapper that does ps, looks for unit test like processes, and uses pstack or gdb to print thread stack traces. Also, using `ps -wwf` instead of `ps -wf` ensures output is not cut off.

For security, CircleCI runs with security restrictions on ptrace (/proc/sys/kernel/yama/ptrace_scope = 1), and this change adds a work-around to `InstallStackTraceHandler()` (only used by testing tools) to allow any process from the same user to debug it. (I've also touched >100 files to ensure all the unit tests call this function.)

Test Plan: local manual + temporary infinite loop in a unit test to observe in CircleCI

### CODE REVIEW

**Comment by pdillinger** [comment]
Hmm.

```
Dumping stacks for 2013812...
sh: 1: pstack: not found
Could not attach to process.  If your uid matches the uid of the target
process, check the setting of /proc/sys/kernel/yama/ptrace_scope, or try
again as the root user.  For more details, see /etc/sysctl.d/10-ptrace.conf
ptrace: Operation not permitted.
```

**Review by hx235 — DISMISSED** [review]
Cool! Just for curiosity, I recalled the current ps prints out something like 
```
circleci    7809    5739  0 17:31 pts/1    00:00:00 make V=1 J=32 -j32 check

circleci  596764    7809  0 17:49 pts/1    00:00:00 /usr/bin/bash -c if test "32" != 1                                  \     && (bu

circleci  597110  596764  0 17:49 pts/1    00:00:00 make T= TMPD=/dev/shm/rocksdb.QYWS check_0

circleci  597415  597110  0 17:49 pts/1    00:00:00 /usr/bin/bash -c export TEST_TMPDIR=/dev/shm/rocksdb.QYWS; \ printf '%s\n' ''???

circleci  597423  597415  8 17:49 pts/1    00:00:26 perl build_tools/gnu_parallel -j32 --plain --joblog=LOG --eta --gnu {} >& t/log-

circleci 1702607  597423  0 17:50 pts/1    00:00:00 /usr/bin/bash -c t/run-db_compaction_test-DBCompactionTest.CompactionLimiter >& 

circleci 1702644 1702607  0 17:50 pts/1    00:00:00 /bin/sh t/run-db_compaction_test-DBCompactionTest.CompactionLimiter

circleci 1702703 1702644  0 17:50 pts/1    00:00:00 ./db_compaction_test --gtest_filter=DBCompactionTest.CompactionLimiter

circleci 2305156  597423  0 17:54 pts/1    00:00:00 ps -wf
```

Are we print stack for every process above?

File: util/hash_test.cc [hunk 1]
```diff
@@ -120,6 +120,7 @@ TEST(HashTest, Hash64Misc) {
       for (uint64_t var_seed = 1; var_seed != 0; var_seed <<= 1) {
         EXPECT_NE(here, Hash64(str.data(), size, var_seed));
       }
+      for (;;) { sleep(1); } // FOR TESTING, LINE 123
```
Comment (hx235) [inline util/hash_test.cc:123]: Just a reminder to remove this before landing. Thanks!

**Comment by pdillinger** [comment]
It's working https://app.circleci.com/pipelines/github/facebook/rocksdb/20092/workflows/868a7b26-d835-4f3c-9674-eb87a9ec3d51/jobs/511044:

```
UID          PID    PPID  C STIME TTY          TIME CMD
root          66       8  0 20:28 pts/1    00:00:00 make V=1 J=32 -j32 check
root       19593      66  0 20:36 pts/1    00:00:00 /usr/bin/bash -c if test "32" != 1                                  \     && (build_tools/gnu_parallel --gnu --help 2>/dev/null) |                    \         grep -q 'GNU Parallel';                                 \ then                                                            \     make T="$t" check_0;                       \ else                                                            \     for t in  cache_test  cache_reservation_manager_test  lru_cache_test  compressed_secondary_cache_test  blob_counting_iterator_test  blob_file_addition_test  blob_file_builder_test  blob_file_cache_test  blob_file_garbage_test  blob_file_reader_test  blob_garbage_meter_test  blob_source_test  db_blob_basic_test  db_blob_compaction_test  db_blob_corruption_test  db_blob_index_test  column_family_test  compact_files_test  clipping_iterator_test  compaction_iterator_test  compaction_job_test  compaction_job_stats_test  compaction_picker_test  compaction_service_test  tiered_compaction_test  comparator_db_test  corruption_test  cuckoo_table_db_test  db_basic_test  db_block_cache_test  db_bloom_filter_test  db_compaction_filter_test  db_compaction_test  db_dynamic_level_test  db_encryption_test  db_flush_test  db_readonly_with_timestamp_test  db_with_timestamp_basic_test  import_column_family_test  db_inplace_update_test  db_io_failure_test  db_iter_test  db_iter_stress_test  db_iterator_test  db_kv_checksum_test  db_log_iter_test  db_memtable_test  db_merge_operator_test  db_merge_operand_test  db_options_test  db_properties_test  db_range_del_test  db_rate_limiter_test  db_secondary_test  db_sst_test  db_statistics_test  db_table_properties_test  db_tailing_iter_test  db_test  db_test2  db_logical_block_size_cache_test  db_universal_compaction_test  db_wal_test  db_with_timestamp_compaction_test  db_write_buffer_manager_test  db_write_test  dbformat_test  deletefile_test  error_handler_fs_test  external_sst_file_basic_test  external_sst_file_test  fault_injection_test  file_indexer_test  filename_test  flush_job_test  listener_test  log_test  manual_compaction_test  memtable_list_test  merge_helper_test  merge_test  obsolete_files_test  options_file_test  perf_context_test  periodic_task_scheduler_test  plain_table_db_test  prefix_test  repair_test  range_del_aggregator_test  range_tombstone_fragmenter_test  seqno_time_test  table_properties_collector_test  version_builder_test  version_edit_test  version_set_test  wal_manager_test  db_wide_basic_test  wide_column_serialization_test  write_batch_test  write_callback_test  write_controller_test  env_basic_test  env_test  io_posix_test  mock_env_test  delete_scheduler_test  prefetch_test  random_access_file_reader_test  auto_roll_logger_test  env_logger_test  event_logger_test  arena_test  memory_allocator_test  inlineskiplist_test  skiplist_test  write_buffer_manager_test  histogram_test  iostats_context_test  statistics_test  stats_history_test  configurable_test  customizable_test  options_settable_test  options_test  block_based_table_reader_test  block_test  data_block_hash_index_test  full_filter_block_test  partitioned_filter_block_test  cleanable_test  cuckoo_table_builder_test  cuckoo_table_reader_test  merger_test  sst_file_reader_test  table_test  block_fetcher_test  testutil_test  block_cache_trace_analyzer_test  io_tracer_parser_test  ldb_cmd_test  reduce_levels_test  sst_dump_test  trace_analyzer_test  block_cache_tracer_test  io_tracer_test  autovector_test  bloom_test  coding_test  crc32c_test  defer_test  dynamic_bloom_test  filelock_test  file_reader_writer_test  hash_test  heap_test  random_test  rate_limiter_test  repeatable_thread_test  ribbon_test  slice_test  slice_transform_test  timer_queue_test  timer_test  thread_list_test  thread_local_test  work_queue_test  agg_merge_test  backup_engine_test  blob_db_test  cassandra_format_test  cassandra_functional_test  cassandra_row_merge_test  cassandra_serialize_test  checkpoint_test  env_timed_test  memory_test  stringappend_test  object_registry_test  option_change_migration_test  options_util_test  hash_table_test  persistent_cache_test  cache_simulator_test  sim_cache_test  compact_on_deletion_collector_test  optimistic_transaction_test  range_locking_test  transaction_test  point_lock_manager_test  write_prepared_transaction_test  write_unprepared_transaction_test  write_committed_transaction_ts_test  timestamped_snapshot_test  ttl_test  util_merge_operators_test  write_batch_with_index_test  c_test; do                                       \       echo "===== Running $t (`date`)"; ./$t || exit 1; done;          \ fi
root       19599   19593  0 20:36 pts/1    00:00:00 make T= check_0
root       19923   19599  0 20:36 pts/1    00:00:00 /usr/bin/bash -c printf '%s\n' ''......\   'To monitor subtest <duration,pass/fail,name>,'..\   '  run "make watch-log" in a separate window' '';..\ { \ .printf './%s\n' db_bloom_filter_test deletefile_test env_test c_test; \ .find t -name 'run-*' -print; \ } \   | perl -pe 's,(^.*MySQLStyleTransactionTest.*$|^.*SnapshotConcurrentAccessTest.*$|^.*SeqAdvanceConcurrentTest.*$|^t/run-table_test-HarnessTest.Randomized$|^t/run-db_test-.*(?:FileCreationRandomFailure|EncodeDecompressedBlockSizeTest)$|^.*RecoverFromCorruptedWALWithoutFlush$),100 $1,' | sort -k1,1gr | sed 's/^[.0-9]* //'....\   | grep -E '.'.....\   | grep -E -v '"^$"'.....\   | build_tools/gnu_parallel -j32 --plain --joblog=LOG --eta --gnu \     --tmpdir=/dev/shm/rocksdb.QRPf '{} >& t/log-{/} || bash -c "cat t/log-{/}; exit $?"' ; \ parallel_retcode=$? ; \ awk '{ if ($7 != 0 || $8 != 0) { if ($7 == "Exitval") { h = $0; } else { if (!f) print h; print; f = 1 } } } END { if(f) exit 1; }' < LOG ; \ awk_retcode=$?; \ if [ $parallel_retcode -ne 0 ] || [ $awk_retcode -ne 0 ] ; then exit 1 ; fi
root       19931   19923 16 20:36 pts/1    00:00:46 perl build_tools/gnu_parallel -j32 --plain --joblog=LOG --eta --gnu --tmpdir=/dev/shm/rocksdb.QRPf {} >& t/log-{/} || bash -c "cat t/log-{/}; exit $?"
root     2019089   19931  0 20:39 pts/1    00:00:00 /usr/bin/bash -c t/run-hash_test-HashTest.Hash64Misc >& t/log-run-hash_test-HashTest.Hash64Misc || bash -c "cat t/log-run-hash_test-HashTest.Hash64Misc; exit $?"
root     2019108 2019089  0 20:39 pts/1    00:00:00 /bin/sh t/run-hash_test-HashTest.Hash64Misc
root     2019112 2019108  0 20:39 pts/1    00:00:00 ./hash_test --gtest_filter=HashTest.Hash64Misc
Dumping stacks for 2019112...
sh: 1: pstack: not found
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f96acaad1b4 in __GI___clock_nanosleep (clock_id=<optimized out>, clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x7ffc340924f0, rem=rem@entry=0x7ffc340924f0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
78	../sysdeps/unix/sysv/linux/clock_nanosleep.c: No such file or directory.

Thread 1 (Thread 0x7f96ac87c740 (LWP 2019112) "hash_test"):
#0  0x00007f96acaad1b4 in __GI___clock_nanosleep (clock_id=<optimized out>, clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x7ffc340924f0, rem=rem@entry=0x7ffc340924f0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007f96acab2ec7 in __GI___nanosleep (requested_time=requested_time@entry=0x7ffc340924f0, remaining=remaining@entry=0x7ffc340924f0) at nanosleep.c:27
#2  0x00007f96acab2dfe in __sleep (seconds=0, seconds@entry=1) at ../sysdeps/posix/sleep.c:55
#3  0x000055b39d02c0bc in HashTest_Hash64Misc_Test::TestBody (this=<optimized out>) at util/hash_test.cc:123
#4  0x000055b39d097d34 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (location=0x55b39d0a6836 "the test body", method=<optimized out>, object=0x55b39e5faa90) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3899
#5  testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object@entry=0x55b39e5faa90, method=<optimized out>, location=location@entry=0x55b39d0a6836 "the test body") at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3935
#6  0x000055b39d08ad66 in testing::Test::Run (this=this@entry=0x55b39e5faa90) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3973
#7  0x000055b39d08b025 in testing::Test::Run (this=0x55b39e5faa90) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3965
#8  testing::TestInfo::Run (this=0x55b39e5f94d0) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4149
#9  0x000055b39d08b1d4 in testing::TestInfo::Run (this=<optimized out>) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4124
#10 testing::TestCase::Run (this=0x55b39e5f9330) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4267
#11 0x000055b39d08b87b in testing::TestCase::Run (this=<optimized out>) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4253
#12 testing::internal::UnitTestImpl::RunAllTests (this=0x55b39e5f9060) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6633
#13 0x000055b39d0982e4 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (location=0x55b39d0a7b50 "auxiliary test code (environments or event listeners)", method=<optimized out>, object=0x55b39e5f9060) at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3899
#14 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x55b39e5f9060, method=<optimized out>, location=location@entry=0x55b39d0a7b50 "auxiliary test code (environments or event listeners)") at third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3935
#15 0x000055b39d08ba33 in testing::UnitTest::Run (this=0x55b39d0bf460 <testing::UnitTest::GetInstance()::instance>) at third-party/gtest-1.8.1/fused-src/gtest/gtest.h:21109
#16 0x000055b39d027fd0 in RUN_ALL_TESTS () at third-party/gtest-1.8.1/fused-src/gtest/gtest.h:22110
#17 main (argc=<optimized out>, argv=0x7ffc340929b8) at util/hash_test.cc:853
[Inferior 1 (process 2019112) detached]
root     2349921   19931  0 20:41 pts/1    00:00:00 sh -c build_tools/ps_with_stack || ps -wwf
root     2349922 2349921  0 20:41 pts/1    00:00:00 perl build_tools/ps_with_stack
root     2349923 2349922  0 20:41 pts/1    00:00:00 ps -wwf
```

> Are we print stack for every process above?

It's matching a command invoked via relative path, such as `./blah -whatever` or `subdir/script param` but not `ps -wwf` nor `/usr/bin/bash -c blah`. A heuristic that should work.

File: port/stack_trace.cc [hunk 2]
```diff
@@ -35,6 +35,9 @@ void* SaveStack(int* /*num_frames*/, int /*first_frames_to_skip*/) {
 #if defined(OS_FREEBSD)
 #include <sys/sysctl.h>
 #endif
+#ifdef OS_LINUX
```
Comment (pdillinger) [inline port/stack_trace.cc:38]: File with important changes

File: port/stack_trace.h [hunk 3]
```diff
@@ -12,7 +12,10 @@ namespace port {
 
 // Install a signal handler to print callstack on the following signals:
 // SIGILL SIGSEGV SIGBUS SIGABRT
-// Currently supports linux only. No-op otherwise.
+// And also (Linux ony for now) overrides security settings to allow outside
```
Comment (pdillinger) [inline port/stack_trace.h:15]: File with comment changes

File: test_util/testharness.h [hunk 4]
```diff
@@ -52,6 +52,8 @@
   } while (false) /* user ; */
 
 #include <string>
+
+#include "port/stack_trace.h"
```
Comment (pdillinger) [inline test_util/testharness.h:56]: So that we don't need to `#include` it in every test file

File: util/heap_test.cc [hunk 5]
```diff
@@ -3,15 +3,16 @@
 //  COPYING file in the root directory) and Apache 2.0 License
 //  (found in the LICENSE.Apache file in the root directory).
 
+#include "util/heap.h"
```
Comment (pdillinger) [inline util/heap_test.cc:6]: From `make format`

File: util/work_queue_test.cc [hunk 6]
```diff
@@ -14,12 +14,15 @@
 #include "util/work_queue.h"
 
 #include <gtest/gtest.h>
+
```
Comment (pdillinger) [inline util/work_queue_test.cc:17]: From `make format`

File: build_tools/gnu_parallel [hunk 7]
```diff
@@ -1916,7 +1916,8 @@ sub drain_job_queue {
                 } elsif (not $ps_reported and (time() - $last_progress_time) >= 60) {
                     # No progress in at least 60 seconds: run ps
                     print $Global::original_stderr "\n";
-                    system("ps", "-wf");
+                    my $script_dir = ::dirname($0);
```
Comment (pdillinger) [inline build_tools/gnu_parallel:1919]: File with important change (call new script, or fall back on ps)

File: build_tools/ps_with_stack [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env perl
```
Comment (pdillinger) [inline build_tools/ps_with_stack:1]: New, important file

**Review by pdillinger — COMMENTED** [review]
Some comments to point out the non-trivial changes

**Review by hx235 — APPROVED** [review]
Still LGTM! 

> It's matching a command invoked via relative path, such as ./blah -whatever or subdir/script param but not ps -wwf nor /usr/bin/bash -c blah. A heuristic that should work

Would you mind adding a code comment in build_tools/ps_with_stack just to point out the reason behind matching "command invoked via relative path" is a heuristic to print stack traces on unit test only? Thanks!

**Pull request closed without merging**